### PR TITLE
Cicd windows post merge

### PR DESCRIPTION
**Pull request by liyier90**
Closes https://github.com/aimakerspace/PeekingDuck-Private/issues/16
Closes https://github.com/aimakerspace/PeekingDuck-Private/issues/23
Closes https://github.com/aimakerspace/PeekingDuck/issues/373

**NOTE:** This PR is based on the changes introduced by https://github.com/aimakerspace/PeekingDuck/pull/563. As a result, this should probably only be review after the previous PR has been merged so that changed files and code are reflected accurately.

Due to [hardware limitation](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners#supported-runners-and-hardware-resources) of GitHub Actions runner, Post Merge CICD fails on Windows. Potentially, this will occur to the Linux runner as well when more model nodes are added in the future.

**Changes in this PR:**
- Create model inside the test functions instead of using fixtures -> *When pytest decides to keep fixtures alive until the end, it won't be large model fixtures*
- Runs `mlmodel` related test scripts individually -> *Release reports frequently so it does not get accumulated*
- `coverage -a` is used to combine the coverage info form all the runs

**Caveat:**
Post Merge Checks takes longer now since loading model takes up time.

### CODE REVIEW

File: tests/pipeline/nodes/model/yolov4_license_plate/test_yolov4_license_plate.py [hunk 1]
```diff
@@ -76,8 +78,9 @@ def test_no_lp_image(self, test_no_lp_images, yolo):
         assert len(output["bbox_labels"]) == 0
         assert len(output["bbox_scores"]) == 0
 
-    def test_at_least_one_lp_image(self, test_lp_images, yolo):
+    def test_at_least_one_lp_image(self, test_lp_images, yolo_type):
         test_img = cv2.imread(test_lp_images)
+        yolo = Node(yolo_type)
         output = yolo.run({"img": test_img})
         assert "bboxes" in output
         assert len(output["bboxes"]) != 0
```
Comment (ongtw) [inline tests/pipeline/nodes/model/yolov4_license_plate/test_yolov4_license_plate.py:86]: Update lines 31-39 to use Pathlib?

**Review by ongtw — CHANGES_REQUESTED** [review]
Suggested some minor updates.

Also, noticed `replace_download_weights(*_)` is defined across many tests files: is it possible to define it once, say in `testconf.py`, and import it instead?

File: scripts/run_tests.sh [hunk 2]
```diff
@@ -24,43 +24,79 @@ show_coverage(){
...
-        show_coverage $showCoverage
+        show_coverage "${showCoverage}"
         echo "TEST PASSED!"
     fi
 
     exit 0
 }
 
-echo "Running $selectedTest tests in $test_dir"
+iterate_run_test() {
```
[the remark this answers is not in the document]
Comment (liyier90) [reply scripts/run_tests.sh]: updated according to suggestion

File: tests/pipeline/nodes/model/efficientdet_d04/test_efficientdet.py [hunk 3]
```diff
@@ -144,10 +152,10 @@ def test_efficientdet_postprocess(self, efficientdet_detector):
...
-        test_models["weighted"] = edet(1, weighted_bifpn=False)
-        test_models["detect_quadrangle"] = edet(1, detect_quadrangle=True)
-        test_models["no_separable_conv"] = edet(1, separable_conv=False)
+        test_models["normal"] = model.efficientdet(1)
+        test_models["weighted"] = model.efficientdet(1, weighted_bifpn=False)
+        test_models["detect_quadrangle"] = model.efficientdet(1, detect_quadrangle=True)
+        test_models["no_separable_conv"] = model.efficientdet(1, separable_conv=False)
 
         for key in test_models:
             assert test_models[key] is not None
```
[the remark this answers is not in the document]
Comment (liyier90) [reply tests/pipeline/nodes/model/efficientdet_d04/test_efficientdet.py:151]: updated

File: tests/pipeline/nodes/model/hrnetv1/test_detector.py [hunk 4]
```diff
@@ -55,18 +54,15 @@ def weights_dir(hrnet_config):
...
+    def test_create_model(self, hrnet_config, model_dir, weights_dir):
         """Testing hrnet model instantiation."""
+        if not checker.has_weights(weights_dir, model_dir):
+            downloader.download_weights(
+                weights_dir, hrnet_config["weights"]["blob_file"]
+            )
+
+        hrnet_detector = Detector(hrnet_config, model_dir)
         hrnet_model = hrnet_detector._create_hrnet_model
         assert hrnet_model is not None
```
[the remark this answers is not in the document]
Comment (liyier90) [reply tests/pipeline/nodes/model/hrnetv1/test_detector.py:59]: updated

File: tests/pipeline/nodes/model/hrnetv1/test_hrnet.py [hunk 5]
```diff
@@ -67,12 +60,15 @@ def test_no_human_image(self, test_no_human_images, hrnet):
...
+    ):
+        """Tests HRnet on images with at least one human present. Bbox
         coordinates is set as the entire image.
         """
         test_img = cv2.imread(test_human_images)
         img_h, img_w, _ = test_img.shape
+        hrnet = Node(hrnet_config)
         output = hrnet.run(
             {"img": test_img, "bboxes": np.array([[0, 0, img_w, img_h]])}
         )
```
[the remark this answers is not in the document]
Comment (liyier90) [reply tests/pipeline/nodes/model/hrnetv1/test_hrnet.py:65]: updated

File: tests/pipeline/nodes/model/hrnetv1/test_hrnet_model.py [hunk 6]
```diff
@@ -61,7 +52,7 @@ def test_no_weights(self, hrnet_config):
             "peekingduck.weights_utils.downloader.download_weights",
             wraps=replace_download_weights,
         ), TestCase.assertLogs(
-            "peekingduck.pipeline.nodes.model.yolov4.yolo_model.logger"
+            "peekingduck.pipeline.nodes.model.hrnetv1.hrnet_model.logger"
         ) as captured:
             hrnet = Node(config=hrnet_config)
             # records 0 - 20 records are updates to configs
```
[the remark this answers is not in the document]
Comment (liyier90) [reply tests/pipeline/nodes/model/hrnetv1/test_hrnet_model.py:45]: updated

File: tests/pipeline/nodes/model/movenetv1/test_movenet.py [hunk 7]
```diff
@@ -321,25 +227,32 @@ def test_multi_human(self, multi_person_image, movenet_model_multi, ground_truth
...
     def test_model_type(self, movenet_config):
         movenet_config["model_type"] = "not_movenet"
-        with pytest.raises(ValueError):
-            movenet = Node(movenet_config)
+        with pytest.raises(ValueError) as excinfo:
+            _ = Node(movenet_config)
+        assert str(excinfo.value) == (
+            "model_type must be one of ['singlepose_lightning', "
+            "'singlepose_thunder', 'multipose_lightning']"
+        )
```
[the remark this answers is not in the document]
Comment (liyier90) [reply tests/pipeline/nodes/model/movenetv1/test_movenet.py:254]: reordered according to suggestion

File: tests/pipeline/nodes/model/mtcnnv1/test_mtcnn.py [hunk 8]
```diff
@@ -87,8 +72,11 @@ def test_no_human_face_image(self, test_no_human_images, mtcnn):
...
 
-    def test_return_at_least_one_face_and_one_bbox(self, test_human_images, mtcnn):
+    def test_return_at_least_one_face_and_one_bbox(
+        self, test_human_images, mtcnn_config
+    ):
         test_img = cv2.imread(test_human_images)
+        mtcnn = Node(mtcnn_config)
         output = mtcnn.run({"img": test_img})
         assert "bboxes" in output
         assert output["bboxes"].size != 0
```
[the remark this answers is not in the document]
Comment (liyier90) [reply tests/pipeline/nodes/model/mtcnnv1/test_mtcnn.py:69]: updated

File: tests/pipeline/nodes/model/posenetv1/test_predictor.py [hunk 9]
```diff
@@ -96,17 +86,17 @@ def test_create_image_from_frame(self, posenet_predictor):
...
+        predictor = Predictor(posenet_config, model_dir)
+        posenet_model = predictor._create_posenet_model()
         assert posenet_model is not None, "Model is not created"
-        coords, scores, masks = posenet_predictor._predict_all_poses(
-            posenet_model, frame, "75"
-        )
+        coords, scores, masks = predictor._predict_all_poses(posenet_model, frame, "75")
         assert coords.shape == (1, 17, 2), "Coordinates is of wrong shape"
         assert scores.shape == (1, 17), "Scores is of wrong shape"
         assert masks.shape == (1, 17), "Masks is of wrong shape"
```
[the remark this answers is not in the document]
Comment (liyier90) [reply tests/pipeline/nodes/model/posenetv1/test_predictor.py:93]: updated, took the chance to apply the similar changes to the .npz file in other posenet test files too

File: tests/pipeline/nodes/model/posenetv1/test_posenet.py [hunk 10]
```diff
@@ -94,3 +95,26 @@ def test_different_models(self, posenet_model, person_image):
...
+                captured.records[0].getMessage()
+                == "---no weights detected. proceeding to download...---"
+            )
+            assert "weights downloaded" in captured.records[1].getMessage()
+            assert posenet is not None
+
+    def test_invalid_config_value(self, posenet_bad_config_value):
+        with pytest.raises(ValueError) as excinfo:
+            _ = Node(config=posenet_bad_config_value)
+        assert "must be" in str(excinfo.value)
```
[the remark this answers is not in the document]
Comment (liyier90) [reply tests/pipeline/nodes/model/posenetv1/test_posenet.py:108]: updated

File: tests/pipeline/nodes/model/yolov4/test_yolo.py [hunk 11]
```diff
@@ -99,7 +99,8 @@ def test_no_weights(self, yolo_config):
             assert "weights downloaded" in captured.records[1].getMessage()
             assert yolo is not None
 
-    def test_get_detect_ids(self, yolo):
+    def test_get_detect_ids(self, yolo_type):
+        yolo = Node(yolo_type)
         assert yolo.model.detect_ids == [0]
 
     def test_yolo_model_initialization(self):
```
[the remark this answers is not in the document]
Comment (liyier90) [reply tests/pipeline/nodes/model/yolov4/test_yolo.py:93]: updated

File: tests/pipeline/nodes/model/yolov4_face/test_yolo_face.py [hunk 12]
```diff
@@ -113,7 +106,8 @@ def test_no_weights(self, yolo_config):
             assert "weights downloaded" in captured.records[1].getMessage()
             assert yolo is not None
 
-    def test_get_detect_ids(self, yolo):
+    def test_get_detect_ids(self, yolo_type):
+        yolo = Node(yolo_type)
         assert yolo.model.get_detect_ids() == [0, 1]
 
     def test_model_initialization(self, yolo_config, model_dir):
```
[the remark this answers is not in the document]
Comment (liyier90) [reply tests/pipeline/nodes/model/yolov4_face/test_yolo_face.py:100]: updated

File: tests/pipeline/nodes/model/yolov4_license_plate/test_yolov4_license_plate.py (hunk 1 above)
Comment (liyier90) [reply tests/pipeline/nodes/model/yolov4_license_plate/test_yolov4_license_plate.py:75]: updated

**Comment by liyier90** [comment]
> Also, noticed replace_download_weights(*_) is defined across many tests files: is it possible to define it once, say in testconf.py, and import it instead?

Implemented this as a fixture instead in `conftest.py` similar to the `create_image` fixture. Did not try the import method as I think it requires some form of relative import or `sys.path` manipulation. And also because we already have a way of sharing a callable with all the tests.

**Review by ongtw — APPROVED** [review]
Looks good

**Pull request merged**